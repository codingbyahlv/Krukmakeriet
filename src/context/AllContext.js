import { createContext, useEffect, useState } from "react";
import { collection, getDocs } from "firebase/firestore";
import { auth, db } from "../firebase-config";
import { onAuthStateChanged } from "firebase/auth";

export const AllContext = createContext();

export const AllContextProvider = ({ children }) => {
  const [courseBasket, setCourseBasket] = useState(
    JSON.parse(localStorage.getItem("courseBasket"))
  );
  const [productBasket, setProductBasket] = useState(
    JSON.parse(localStorage.getItem("productBasket"))
  );
  const [adminPermission, setAdminPermission] = useState(false);
  const [refresh, setRefresh] = useState(true);
  const usersRef = collection(db, "users");
  const [shopProductModalOpen, setShopProductModalOpen] = useState(false);
  const [selectedProduct, setSelectedProduct] = useState(null);
  const [showTopBtn, setShowTopBtn] = useState(false);
  const [addedToBasketPopupOpen, setAddedToBasketPopupOpen] = useState(false);
  const [basketAmount, setBasketAmount] = useState(0);

  if (courseBasket === null) {
    setCourseBasket([]);
  }
  if (productBasket === null) {
    setProductBasket([]);
  }

  useEffect(() => {
    localStorage.setItem("courseBasket", JSON.stringify(courseBasket));
  }, [courseBasket]);

  useEffect(() => {
    localStorage.setItem("productBasket", JSON.stringify(productBasket));
  }, [productBasket]);

  useEffect(() => {
    async function getUsers() {
      const data = await getDocs(usersRef);
      const users = data.docs.map((doc) => ({ ...doc.data() }));
      onAuthStateChanged(auth, async (user) => {
        if (!user) return;

        const isAdmin = users.find((a) => a.uid === user.uid);
        setAdminPermission(isAdmin.admin);
        localStorage.setItem("admin", isAdmin.admin);
      });
    }
    getUsers();
  }, [refresh, usersRef]);

  return (
    <AllContext.Provider
      value={{
        courseBasket,
        setCourseBasket,
        productBasket,
        setProductBasket,
        adminPermission,
        setAdminPermission,
        refresh,
        setRefresh,
        shopProductModalOpen,
        setShopProductModalOpen,
        selectedProduct,
        setSelectedProduct,
        showTopBtn,
        setShowTopBtn,
        addedToBasketPopupOpen,
        setAddedToBasketPopupOpen,
        basketAmount,
        setBasketAmount,
      }}
    >
      {children}
    </AllContext.Provider>
  );
};
