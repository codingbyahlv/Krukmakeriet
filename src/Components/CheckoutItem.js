import React, { useContext } from "react";
import styles from "./CheckoutItem.module.css";
import { AllContext } from "../context/AllContext";
import { FaTrash } from "react-icons/fa";

function CheckoutItem({ productData }) {
  const { productBasket, setProductBasket, courseBasket, setCourseBasket } =
    useContext(AllContext);

  const basketZero = productData.amount === 0;

  const isProduct = courseBasket.some((product) => {
    if (product.id === productData.id) {
      return true;
    }
    return false;
  });

  function decrementAmount(productData) {
    const courseExist = courseBasket.find((item) => item.id === productData.id);
    const productExist = productBasket.find(
      (item) => item.id === productData.id
    );

    if (isProduct) {
      courseExist.amount > 0
        ? setCourseBasket(
            courseBasket.map((item) =>
              item.id === productData.id
                ? { ...courseExist, amount: courseExist.amount + -1 }
                : item
            )
          )
        : console.log("Nothing to remove");
    } else {
      productExist.amount > 0
        ? setProductBasket(
            productBasket.map((item) =>
              item.id === productData.id
                ? { ...productExist, amount: productExist.amount + -1 }
                : item
            )
          )
        : console.log("Nothing to remove");
    }
  }

  function incrementAmount(productData) {
    const productExist = productBasket.find(
      (item) => item.id === productData.id
    );
    setProductBasket(
      productBasket.map((item) =>
        item.id === productData.id
          ? { ...productExist, amount: productExist.amount + 1 }
          : item
      )
    );

    const courseExist = courseBasket.find((item) => item.id === productData.id);
    setCourseBasket(
      courseBasket.map((item) =>
        item.id === productData.id
          ? { ...courseExist, amount: courseExist.amount + 1 }
          : item
      )
    );
  }

  function deleteItem(productData) {
    setCourseBasket(courseBasket.filter((item) => item.id !== productData.id));
    setProductBasket(
      productBasket.filter((item) => item.id !== productData.id)
    );
  }

  return (
    <article className={styles.wrapper} data-testid="checkoutItem">
      <div className={styles.imgWrapper}>
        <img src={productData.img} alt="" className={styles.basketImage} />
      </div>

      <div className={styles.textWrapper}>
        <p className={styles.heading}>{productData.name}</p>
        <div className={styles.changeAmountWrapper}>
          {basketZero ? (
            <p
              className={styles.changeAmount}
              onClick={() => decrementAmount(productData)}
            ></p>
          ) : (
            <p
              className={styles.changeAmount}
              onClick={() => decrementAmount(productData)}
            >
              {" "}
              -{" "}
            </p>
          )}

          <p>{productData.amount}</p>
          <p
            className={styles.changeAmount}
            data-testid="incrementBtn"
            onClick={() => incrementAmount(productData)}
          >
            +
          </p>
        </div>
      </div>

      <div className={styles.priceDeleteWrapper}>
        <p className={styles.price}>{productData.price}:-</p>
        <FaTrash
          className={styles.deleteBtn}
          data-testid="decrementBtn"
          onClick={() => deleteItem(productData)}
        />
      </div>
    </article>
  );
}

export default CheckoutItem;
