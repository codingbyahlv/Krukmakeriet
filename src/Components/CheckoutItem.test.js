import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import CheckoutItem from "./CheckoutItem";
import { AllContextProvider } from "../context/AllContext";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

const mockProductData = [
  {
    name: "skÃ¥l",
    id: "test123",
    img: "https",
  },
];
const MockCheckoutItem = () => {
  return (
    <BrowserRouter>
      <AllContextProvider>
        <CheckoutItem productData={mockProductData} />
      </AllContextProvider>
    </BrowserRouter>
  );
};

describe("CheckoutItem", () => {
  test("check if props exist as should", async () => {
    render(<MockCheckoutItem />);
    const routeForCheckoutItem = screen.getByTestId("checkoutItem");
    expect(routeForCheckoutItem).toBeInTheDocument();
  });
  test("create a snapshot to test the rendering of element", async () => {
    const domTree = renderer
      .create(
        <BrowserRouter>
          <AllContextProvider>
            <CheckoutItem productData={mockProductData} />
          </AllContextProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });
  // test("check if onclick on increment button the amount changes", async () => {
  //   render(<MockCheckoutItem />);
  //   const incrementBtn = screen.getByTestId("incrementBtn");
  //   const counterText = screen.getByText("0");
  //   expect(counterText.child).toHaveTextContent("0");
  //   fireEvent.click(incrementBtn);
  //   console.log();
  //   expect(counterText.textContent).toMatch("1");
  // });

  // test("check if onclick on decrement button the amount changes", async () => {
  //   render(<MockCheckoutItem />);
  //   const decremetnBtn = screen.getByTestId("decrementBtn");
  //   const counterText = screen.getByTestId("counterText");
  //   console.log(counterText.textContent);
  //   expect(counterText).toHaveTextContent("1");
  //   console.log(counterText.textContent);
  //   fireEvent.click(decremetnBtn);
  //   expect(counterText).toHaveTextContent("");
  // });
});
