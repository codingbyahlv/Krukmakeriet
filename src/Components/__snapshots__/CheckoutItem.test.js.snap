// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutItem create a snapshot to test the rendering of element 1`] = `
<article
  className="wrapper"
  data-testid="checkoutItem"
>
  <div
    className="imgWrapper"
  >
    <img
      alt=""
      className="basketImage"
    />
  </div>
  <div
    className="textWrapper"
  >
    <p
      className="heading"
    />
    <div
      className="changeAmountWrapper"
    >
      <p
        className="changeAmount"
        onClick={[Function]}
      >
         
        -
         
      </p>
      <p />
      <p
        className="changeAmount"
        data-testid="incrementBtn"
        onClick={[Function]}
      >
        +
      </p>
    </div>
  </div>
  <div
    className="priceDeleteWrapper"
  >
    <p
      className="price"
    >
      :-
    </p>
    <svg
      className="deleteBtn"
      data-testid="decrementBtn"
      fill="currentColor"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
      />
    </svg>
  </div>
</article>
`;
